src/bosh-director/lib/bosh/director/deployment_plan/steps/rotate_nats_bootstrap_credentials_step.rb

```diff
@@ -0,0 +1,21 @@
+module Bosh::Director
+  module DeploymentPlan
+    module Steps
+      class RotateNatsBootstrapCredentialsStep
+        def perform(report)
+          agent_permanent_creds = NatsClientCertGenerator.new(@logger).generate_nats_client_certificate "#{report.vm.agent_id}.agent.bosh-internal"
+          settings = {}
+          settings['mbus'] = {
+            'cert' => {
+              'ca' => Config.nats_server_ca,
+              'certificate' => agent_permanent_creds[:cert].to_pem,
+              'private_key' => agent_permanent_creds[:key].to_pem,
+            },
+          }
+          AgentClient.with_agent_id(report.vm.agent_id, report.vm.instance.name).update_settings(settings)
```

Comment:
I'm not sure it's safe to perform a partial `update_settings` to the agent here. The agent looks like it expects all the data for `update_settings` and if you don't sent the Trusted Certs, it will blank them out again: https://github.com/cloudfoundry/bosh-agent/blob/6afc0689eeabe7673de54ab0f293df2c746030f1/agent/action/update_settings.go#L63

I think we ran into a similar problem when we were doing the live NATS cert rotation, where `update_settings` was getting called from multiple places, and we had to send all the data every time. At that point we pulled the `update_settings` call into a central place to make it easier to keep them all in sync: https://github.com/cloudfoundry/bosh/blob/3b031a55e4eabbdb6aa7389d9fef444f52aab219/src/bosh-director/lib/bosh/director/deployment_plan/instance.rb#L181

You'll probably need to somehow use that existing tooling (it will likely need to be modified), or you'd have to duplicate the logic and pull the other parts needed into this new Step so it can send the full `update_settings` data too.

Reply:
In fact, if you could eliminate the new Step entirely and push the logic down into the existing `UpdateInstanceSettingsStep`, it might make things easier and cleaner.

Reply:
@jpalermo thanks for noticing the problem, we're trying to replicate and fix it. About moving things to `UpdateInstanceSettingsStep` , it will be difficult to track where the rotation will be triggered, because `instance.update_instance_settings` gets called in other places, so adding more logic there could bring unforeseen behaviors.
This function gets called in the following places according to the IDE:
`recreate_handler.rb:39`
`update_procedure.rb:153`
Do you think it's safe to rotate NATs creds everytime these functions get executed?

Reply:
`UpdateInstanceSettingsStep` is only used during create, so you can assume you want to regenerate the NATS creds if that step is called (and the feature is enable).

`instance.update_instance_settings`, the underlying function, does get called in a couple of places, you'd probably want to change the method signature to take an additional argument telling it you want to force new NATS creds. Might be other ways to accomplish it too.

Adding more logic and branching to `instance.update_instance_settings` would make it more complicated and add risk. But all that logic would at least be in a single place rather than having multiple spots in the codebase where UpdateSettings could be happening.

Reply:
Hi @jpalermo the changes were completed and now the rotation is happening in `UpdateInstanceSettingsStep`.